oscore-cf/src/main/java/org/eclipse/californium/oscore/ObjectSecurityLayer.java

```diff
@@ -0,0 +1,277 @@
+	 * The logger
+	 */
+	private static final Logger LOGGER = LoggerFactory.getLogger(ObjectSecurityLayer.class.getName());
+
+	private static final byte[] EMPTY = new byte[0];
+
+	/**
+	 * Encrypt an outgoing request using the OSCore context.
+	 * 
+	 * @param message the message
+	 * @param ctx the OSCore context
+	 * 
+	 * @return the encrypted message
+	 * 
+	 * @throws OSException error while encrypting request
+	 */
+	public static Request prepareSend(Request message, OSCoreCtx ctx) throws OSException {
+		return RequestEncryptor.encrypt(message, ctx);
+	}
+
+	/**
+	 * Encrypt an outgoing response using the OSCore context.
+	 * 
+	 * @param message the message
+	 * @param ctx the OSCore context
+	 * 
+	 * @return the encrypted message
+	 * 
+	 * @throws OSException error while encrypting response
+	 */
+	public static Response prepareSend(Response message, OSCoreCtx ctx, final boolean newPartialIV) throws OSException {
+		return ResponseEncryptor.encrypt(message, ctx, newPartialIV);
+	}
+
+	/**
+	 * Decrypt an incoming request using the right OSCore context
+	 * 
+	 * @param request the incoming request
+	 * @param db the OSCore context database
+	 * 
+	 * @return the cid of the OSCore context used for this request
+	 * 
+	 * @throws OSException error while decrypting request
+	 */
+	public static Request prepareReceive(Request request) throws CoapOSException {
+		return RequestDecryptor.decrypt(request);
+	}
+
+	/**
+	 * Decrypt an incoming response using the right OSCore context
+	 * 
+	 * @param response the incoming request
+	 * @param db the OSCore context database
+	 * @return the decrypted and verified response
+	 * 
+	 * @throws OSException error while decrypting response
+	 */
+	public static Response prepareReceive(Response response) throws OSException {
+		return ResponseDecryptor.decrypt(response);
+	}
+
+	@Override
+	public void sendRequest(Exchange exchange, Request request) {
+		Request req = request;
+		if (shouldProtectRequest(request)) {
+			try {
+				String uri = request.getURI();
+				final OSCoreCtxDB db = HashMapCtxDB.getInstance();
+
+				if (uri == null) {
+					LOGGER.error(Error.URI_NULL);
+					throw new OSException(Error.URI_NULL);
+				}
+				if (db == null) {
+					LOGGER.error(Error.DB_NULL);
+					throw new OSException(Error.DB_NULL);
+				}
+
+				final OSCoreCtx ctx = db.getContext(uri);
+				if (ctx == null) {
+					LOGGER.error(Error.CTX_NULL);
+					throw new OSException(Error.CTX_NULL);
+				}
+
+				exchange.setCryptographicContextID(ctx.getRecipientId());
+				final int seqByToken = ctx.getSenderSeq();
+
+				final Request preparedRequest = prepareSend(request, ctx);
+				preparedRequest.addMessageObserver(new MessageObserverAdapter() {
+
+					@Override
+					public void onReadyToSend() {
+						Token token = preparedRequest.getToken();
+						db.addContext(token, ctx);
+						db.addSeqByToken(token, seqByToken);
+					}
+				});
+
+				req = preparedRequest;
+
+				if (req.getToken() != null) {
```

Comment:
This will be now added in the `onReadyToSend` callback.
Please remove lines 142-145.

Reply:
My mistake.